tests/TypeSpecifierOrContextsTest.hack

```diff
@@ -0,0 +1,51 @@
+
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+
+namespace Facebook\HHAST;
+
+use function Facebook\FBExpect\expect;
+use type Facebook\HackTest\DataProvider;
+
+final class TypeSpecifierOrContextsTest extends TestCase {
+  public function getExamples(): dict<string, (string, classname<Node>)> {
+    return dict[
+      'unwrapped type specifier' => tuple(
+        'abstract const type TFoo as shape(...)',
+        ShapeTypeSpecifier::class,
+      ),
+      'wrapped type specifier' => tuple(
+        'abstract const type TFoo as MyClass',
+        SimpleTypeSpecifier::class,
+      ),
+      'contexts' =>
+        tuple('abstract const ctx CFoo super [defaults]', Node::class),
```

Comment:
should this be `Contexts::class`?

Reply:
whoops yep, but the SimpleTypeSpecifier one is the surprise

Reply:
> don't forget to add ContextConstDeclaration to IClassBodyDeclaration

Done :)

> do we want a similar interface (ITypeSpecifierOrVariableToken?) for the other one too?


I don't think so: VariableTokens actually are *sometimes* specifying a type, similar to how a NameToken is *sometimes* specifying a type, depending on context. If it's not, you get a raw NameToken, if it is we generate a SimpleTypeSpecifier that doesn't exist in the CST